src/filesystem.ts

```diff
@@ -105,6 +113,70 @@ export async function loadFileSystem({ config, dependencies, rootKey, username }
   return fs
 }
 
+/**
+ * Recover a user's file system.
+ */
+export async function recoverFileSystem({
+  agentDID,
+  auth,
+  dependencies,
+  oldUsername,
+  newUsername,
+  readKey,
+}: {
+  agentDID: () => Promise<string>
+  auth: AuthenticationStrategy
+  dependencies: Dependencies
+} & RecoverFileSystemParams): Promise<{ success: boolean }> {
+  try {
+    const { crypto, reference, storage } = dependencies
+
+    const newRootDID = await agentDID()
+
+    // Register a new user with the `newUsername`
+    const { success } = await auth.register({
+      username: newUsername,
+    })
+    if (!success) {
+      throw new Error("Failed to register new user")
+    }
+
+    // Build an ephemeral UCAN to authorize the dataRoot.update call
+    const proof: string | null = await storage.getItem(storage.KEYS.ACCOUNT_UCAN)
+    const ucan = await Ucan.build({
+      dependencies,
+      potency: "APPEND",
+      resource: "*",
+      proof: proof ? proof : undefined,
+      lifetimeInSeconds: 60 * 3, // Three minutes
+      audience: newRootDID,
+      issuer: newRootDID,
+    })
+
+    const oldRootCID = await reference.dataRoot.lookup(oldUsername)
+    if (!oldRootCID) {
+      throw new Error("Failed to lookup oldUsername")
+    }
+
+    // Update the dataRoot of the new user
+    await reference.dataRoot.update(oldRootCID, ucan)
+
+    // Store the accountDID, which is used to namespace the readKey
```

Comment:
Might be more correct to say: "Store the read key which namespaced using the account DID"?